examples/cis2-wccd-upgradable/src/lib.rs

```diff
@@ -0,0 +1,3512 @@
+    /// Contract is paused/unpaused.
+    paused:             bool,
+}
+
+#[derive(Serialize, PartialEq)]
+enum ProtocolAddressesState {
+    UnInitialized,
+    Initialized {
+        /// Address of the w_ccd proxy contract.
+        proxy_address:          ContractAddress,
+        /// Address of the w_ccd implementation contract.
+        implementation_address: ContractAddress,
+    },
+}
+
+/// The `implementation` contract state.
+#[derive(Serial, Deserial, SchemaType)]
+struct StateImplementation {
+    /// The admin address can pause/unpause the contract
+    /// and set implementors.
+    admin:              Address,
+    /// Addresses of the protocol
+    protocol_addresses: ProtocolAddressesImplementation,
+}
+
+#[derive(SchemaType, Serialize, PartialEq)]
+enum ProtocolAddressesImplementation {
+    UnInitialized,
+    Initialized {
+        /// Address of the w_ccd proxy contract.
+        proxy_address: ContractAddress,
+        /// Address of the w_ccd state contract.
+        state_address: ContractAddress,
+    },
+}
+
+/// The `proxy` contract state.
+#[derive(Serial, Deserial, SchemaType)]
+struct StateProxy {
+    /// The admin address can upgrade the implementation contract.
+    admin:                  Address,
+    /// Address of the w_ccd implementation contract.
+    implementation_address: ContractAddress,
+    /// Address of the w_ccd state contract.
+    state_address:          ContractAddress,
+}
+
+/// The mint event parameter type for the contract function `logEvent`.
+#[derive(Serialize)]
+struct MintEventParams {
+    /// The amount of tokens to be minted.
+    amount:  ContractTokenAmount,
+    // Address to receive the minted tokens.
+    address: Address,
+}
+
+/// The burn event parameter type for the contract function `logEvent`.
+#[derive(Serialize)]
+struct BurnEventParams {
+    /// The amount of tokens to be burned.
+    amount:  ContractTokenAmount,
+    // Address that the tokens be burned from.
+    address: Address,
+}
+
+/// The transfer parameter type for the contract function `logEvent`.
+#[derive(Serialize)]
+struct TransferEventParams {
+    /// The amount of tokens to be transferred.
+    amount: ContractTokenAmount,
+    /// Address that the tokens be transferred from.
+    from:   Address,
+    /// Address to receive the tokens.
+    to:     Address,
+}
+
+/// The update operator parameter type for the contract function `logEvent`.
+#[derive(Serialize)]
+struct UpdateOperatorEventParams {
+    /// The owner of the tokens.
+    owner:    Address,
+    /// The operator of the tokens.
+    operator: Address,
+    /// Add or remove an operator.
+    update:   OperatorUpdate,
+}
+
+/// NewAdminEvent.
+#[derive(Serial)]
+struct NewAdminEvent {
+    /// New admin address.
+    new_admin: Address,
+}
+
+/// NewImplementationEvent.
+#[derive(Serial)]
+struct NewImplementationEvent {
+    /// New implementation address.
+    new_implementation: ContractAddress,
+}
```

Comment:
You probably want to merge these types into an enum, since this would put a tag byte in front of the serialization.
Now, without the tag, it is difficult to distinguish the events provided just the bytes.